File: demos/scripts/tour.py
"""Tour demo: short mouse-driven sweep across every plugin surface.

No chat question (the per-token streaming latency made the v1 tour
drag); every beat is a mouse-click that exposes a different surface,
held just long enough to be recognisable.

Beats (each ~3-4 s):
  1. Collapsed-sidebar baseline so the status bar + ribbons read.
  2. Ribbon -> open chat sidebar.
  3. Ribbon -> open Task Center sidebar.
  4. Cmd-id -> open Catalog modal.
  5. Cmd-id -> open Documents modal.
  6. Cmd-id -> open Wiki sidebar.
  7. Cmd-id -> open Settings, click the lilbee tab.
  8. Close back to chat.
"""
from __future__ import annotations

from _mouse import click_locator
from _record import jitter_sleep
from _setup import prepare
from playwright.sync_api import Page


def run(page: Page) -> None:
    prepare(page)

    # Baseline: sidebars collapsed, lilbee leaves detached.
    page.evaluate('''() => {
        const app = window.app;
        app.workspace.detachLeavesOfType('lilbee-tasks');
        app.workspace.detachLeavesOfType('lilbee-wiki');
        if (app.workspace.leftSplit && !app.workspace.leftSplit.collapsed) app.workspace.leftSplit.collapse();
        if (app.workspace.rightSplit && !app.workspace.rightSplit.collapsed) app.workspace.rightSplit.collapse();
    }''')
    jitter_sleep(1.5)

    # Beat 2: Ribbon -> chat.
    click_locator(page, page.locator('[aria-label="Open lilbee chat"]').first, duration=0.5)
    jitter_sleep(2.5)

    # Beat 3: Ribbon -> Task Center.
    click_locator(page, page.locator('[aria-label="Open lilbee Task Center"]').first, duration=0.5)
    jitter_sleep(2.5)

    # Beat 4: Catalog modal.
    page.evaluate('() => window.app.commands.executeCommandById("lilbee:lilbee:catalog")')
    jitter_sleep(2.5)
    page.keyboard.press("Escape")
    jitter_sleep(0.5)

    # Beat 5: Documents modal.
    page.evaluate('() => window.app.commands.executeCommandById("lilbee:lilbee:documents")')
    jitter_sleep(2.5)
    page.keyboard.press("Escape")
    jitter_sleep(0.5)

    # Beat 6: Wiki sidebar.
    page.evaluate('() => window.app.commands.executeCommandById("lilbee:lilbee:wiki")')
    jitter_sleep(2.5)

    # Beat 7: Settings + click the lilbee tab.
    page.evaluate('() => window.app.commands.executeCommandById("app:open-settings")')
    jitter_sleep(1.2)
    lilbee_tab = page.locator('.vertical-tab-nav-item:has-text("lilbee")').first
    lilbee_tab.scroll_into_view_if_needed()
    click_locator(page, lilbee_tab, duration=0.4)
    jitter_sleep(2.5)
    page.keyboard.press("Escape")
    jitter_sleep(0.5)

    # Beat 8: closing frame on chat.
    page.evaluate('''() => {
        const app = window.app;
        const chatLeaf = app.workspace.getLeavesOfType('lilbee-chat')[0];
        if (chatLeaf) app.workspace.revealLeaf(chatLeaf);
    }''')
    jitter_sleep(2.0)

File: demos-src/tapes/tour.ts
/**
 * command_palette demo: the palette is the async control surface for lilbee.
 *
 * Fire three long-running jobs back to back, each through the visible palette
 * flow, without waiting for any of them to finish:
 *
 *   1. Add the open file         (Add current file)
 *   2. Crawl a web page          (Crawl web page + URL)
 *   3. Download a model          (Browse model catalog + pull featured Qwen3 0.6B)
 *
 * Add fires first so its task is enqueued before the crawl's sync covers the
 * same vault file and turns the add into a server-side no-op.
 *
 * The Task Center then shows all three running at once — a crawl, an ingest,
 * and a download in parallel — which is the point: the GUI never blocks. Once
 * the crawl and ingest settle, ask the just-crawled page a question for a
 * cited answer while the download keeps running behind the chat.
 *
 * Recorded with the browser cache wiped, so the crawl's first-run Chromium
 * setup is one of the visible Task Center jobs.
 */
import {
  beat,
  clickSelector,
  clickSend,
  fillChat,
  key,
  runJs,
  sleep,
  storyboard,
  type_,
  waitForSelector,
} from "../src/lib.ts";

const CRAWL_URL = "https://en.wikipedia.org/wiki/Knowledge_graph";
const ADD_FILE = "Notes/Crown Vic upgrade log.md";
const QUESTION = "What is a knowledge graph?";

// Featured pick on the catalog's opening screen — no search needed.
const MODEL_REPO = "Qwen/Qwen3-0.6B-GGUF";
const MODEL_CARD = `.lilbee-model-card[data-repo="${MODEL_REPO}"]`;

const palette = (label: string, query: string, holdAfter = 1000) => [
  beat(
    `Open the command palette (${label})`,
    runJs(`window.app.commands.executeCommandById("command-palette:open");`),
    { holdMs: 500, keyHint: "⌘P" },
  ),
  beat(`Type "${query}"`, type_(query), { holdMs: 1000 }),
  beat(`Run ${label}`, key("enter"), { holdMs: holdAfter }),
];

export default storyboard("tour", {
  window: [1400, 900],
  layout: "explorer-chat-tasks",
  preloadChatModel: true,
  clearTaskCenter: true,
  clearChat: true,
  // Drop the crawled page and the add target so both are real jobs, and
  // remove the model so the pull is a real download.
  freshIngest: ["index.md", "Crown Vic upgrade log.md"],
  freshModel: MODEL_REPO,
  beats: [
    beat("Opening hold on the chat + Task Center", sleep(300)),

    // Open the file we'll add, beside the chat (never over the Task Center).
    beat(
      "Open a note to add, beside the chat",
      runJs(`
        const app = window.app;
        const chatLeaf = app.workspace.getLeavesOfType('lilbee-chat')[0];
        if (chatLeaf) app.workspace.setActiveLeaf(chatLeaf, { focus: true });
        const file = app.vault.getAbstractFileByPath(${JSON.stringify(ADD_FILE)});
        const leaf = app.workspace.getLeaf('tab');
        if (file) await leaf.openFile(file);
        await new Promise(r => setTimeout(r, 250));
      `),
      { holdMs: 700 },
    ),

    // --- Job 1: add the open file ---
    ...palette("Add current file", "Add current file", 900),

    // --- Job 2: crawl a web page, without waiting for the add ---
    ...palette("Crawl web page", "Crawl web page", 900),
    beat(
      "Paste the URL into the crawl modal",
      runJs(`
        const input = document.querySelector('input.lilbee-crawl-url');
        if (input) {
          const setter = Object.getOwnPropertyDescriptor(HTMLInputElement.prototype, 'value').set;
          setter.call(input, ${JSON.stringify(CRAWL_URL)});
          input.dispatchEvent(new Event('input', { bubbles: true }));
        }
      `),
      { holdMs: 900 },
    ),
    beat("Click Crawl", clickSelector('.modal-container button.mod-cta:has-text("Crawl")'), { holdMs: 1000 }),

    // --- Job 3: download a model, again without waiting ---
    ...palette("Browse model catalog", "Browse model catalog", 1000),
    beat("Wait for the featured picks", waitForSelector(MODEL_CARD), { holdMs: 800 }),
    beat("Click Download on the featured card", clickSelector(`${MODEL_CARD} .lilbee-catalog-pull`), { holdMs: 1000 }),
    beat("Confirm the download", clickSelector(".lilbee-confirm-pull-actions button.mod-cta"), { holdMs: 900 }),
    beat("Close the catalog so the Task Center is in view", key("escape"), { holdMs: 1000 }),

    // The money shot: an ingest, a crawl, and a download all running at once.
    // Wait only for what the chat needs — the crawled page searchable and the
    // non-pull jobs idle for a few polls. No first-seen-activity guard: on a
    // fast network the jobs can finish before this beat starts, and the beat
    // must then move straight on. Hard cap keeps a pathological wait off the
    // reel. The download keeps running while we chat, which is the point.
    beat(
      "Three jobs run in parallel in the Task Center",
      runJs(`
        const plugin = window.app.plugins.plugins.lilbee;
        const tq = plugin.taskQueue;
        const busyNonPull = () =>
          [...tq.activeAll, ...tq.queued].filter(t => t.type !== 'pull').length;
        const crawledIngested = async () => {
          try {
            const res = await plugin.api.listDocuments('Knowledge_graph');
            return (res.documents ?? []).length > 0;
          } catch { return false; }
        };
        let quiet = 0;
        for (let i = 0; i < 240; i++) {
          quiet = busyNonPull() > 0 ? 0 : quiet + 1;
          if (quiet >= 4 && await crawledIngested()) break;
          await new Promise(r => setTimeout(r, 500));
        }
        // On a fast network the pull wins the race and its auto-activate
        // switches the chat model; restore the daily-driver pick so the chat
        // scene matches the opening rail. Happens inside the 10x window.
        const pullBusy = [...tq.activeAll, ...tq.queued].some(t => t.type === 'pull');
        const daily = 'Qwen/Qwen3-8B-GGUF/Qwen3-8B-Q4_K_M.gguf';
        if (!pullBusy && plugin.activeModel !== daily) {
          await plugin.api.setChatModel(daily);
          await plugin.fetchActiveModel();
          plugin.refreshOpenChatRails();
          await new Promise(r => setTimeout(r, 400));
        }
      `),
      { holdMs: 1600, speedup: 10, maxMs: 600_000 },
    ),

    // Now use the page the crawl just fetched: a cited answer.
    beat(
      "Activate a clean chat panel",
      runJs(`
        const leaves = window.app.workspace.getLeavesOfType('lilbee-chat');
        if (leaves[0]) window.app.workspace.revealLeaf(leaves[0]);
        await new Promise(r => setTimeout(r, 300));
        const ta = document.querySelector('textarea.lilbee-chat-textarea');
        if (ta) ta.focus();
      `),
      { holdMs: 500 },
    ),
    beat("Ask about the knowledge graph", fillChat(QUESTION), { holdMs: 600 }),
    beat(
      "Ensure the question is in the box",
      runJs(`
        const ta = document.querySelector('textarea.lilbee-chat-textarea');
        if (ta && !ta.value.trim()) {
          const setter = Object.getOwnPropertyDescriptor(HTMLTextAreaElement.prototype, 'value').set;
          setter.call(ta, ${JSON.stringify(QUESTION)});
          ta.dispatchEvent(new Event('input', { bubbles: true }));
        }
      `),
      { holdMs: 300 },
    ),
    beat("Send", clickSend(), { holdMs: 600 }),
    beat(
      "Cited answer from the just-crawled page",
      runJs(`
        const send = document.querySelector('.lilbee-chat-send');
        for (let i = 0; i < 240; i++) {
          const t = (send?.textContent || '').toLowerCase();
          if (t.includes('send') && i > 4) return;
          await new Promise(r => setTimeout(r, 500));
        }
      `),
      { holdMs: 1400, speedup: 4, maxMs: 180_000 },
    ),
    beat(
      "Expand sources",
      runJs(`document.querySelectorAll('.lilbee-chat-sources details').forEach(d => d.open = true);`),
      { holdMs: 1800 },
    ),
  ],
});
